### ISSUE TITLE
Markdown data source

### ISSUE THREAD

**Issue by mholt** [timelinize/timelinize#12]
Someone on Discord mentioned they keep a journal as Markdown files.

A markdown importer seems like a good idea in general. We'll need to figure out how to extract date information from it.

Sometimes Markdown files contain "front matter", or a kind of a header, that might include dates. We could parse front matter (Caddy does this already in its template handler, and can be used as a starting point).

**Comment by rubiojr** [comment]
This would be nice, also as a way to store random text notes in our timelines.

**Comment by mholt** [comment]
Yes, that's one option! Although I have some thoughts on timeline "curation" as I call it (manually adding notes or any kind of data to the timeline) that might be separate from a Markdown data source.

In a branch that I haven't pushed yet, I am building a feature that allows you to write "stories" which are basically like rich text docs with real data from your timeline embedded. Think like... a formatted text document with a WYSIWYG editor (TinyMCE), where you can search for and drag timeline content into your document. So like a journal feature, but with live, dynamic views of the data you're talking about. It could be a whole group of pictures, a track of locations, or parts of a conversation, embedded right into your document. It can even be a custom DB query! You can see the groundwork is already laid in the schema:

https://github.com/timelinize/timelinize/blob/e296b73d2da58124a1a9e3dc9759b3f5da9d8ede/timeline/schema.sql#L272-L300

I also have plans to be able to design PDFs that contain your timeline data, suitable for printing or sharing. Think like a photo book, but with more than just photos.

The schema also supports adding notes directly to items:

https://github.com/timelinize/timelinize/blob/e296b73d2da58124a1a9e3dc9759b3f5da9d8ede/timeline/schema.sql#L194

And we can add a "note" field to more tables if we want (like entities, relationships, attributes, etc).

Of course, if you want to add actual data points to your timeline manually, that can be done too. I have another branch where I've started working on manual data entry and interactive imports.

**Comment by rubiojr** [comment]
The feature you are describing and working on would definitely cover my needs. Annotating timeline items and make those searchable would be fantastic.

**Comment by rubiojr** [comment]
Happy to contribute this one. Frontend is one of my weaknesses, so I can't promise I'll come back with a UI for rendering this soon 😃.

**Comment by mholt** [comment]
I can help with the UI rendering!

**Comment by mholt** [comment]
@rubiojr I've implemented (very rudimentary) support for rendering Markdown, text, and HTML items in [8f0f491](https://github.com/timelinize/timelinize/commit/8f0f491ad5ee322b1974388c1a584e49c060f8c5):

![image](https://github.com/user-attachments/assets/a5462ef1-54d1-40ce-bb9e-e47278194635)

If the item's class is "document" or "note" (not yet implemented in the backend), it will show up like this in the timeline.

**Comment by rubiojr** [comment]
@mholt that looks lovely.

Before giving the backend a go: would you like to see dedicated data sources for them (markdown, html) or would you rather have the generic data source classify them based on file extension?

**Comment by mholt** [comment]
@rubiojr Good question. Maybe the generic data source can do it for now. There's not much special about either type -- _maybe_ getting metadata/dates from these files is "special", but not special enough for their own data source... getting metadata can be a function that the generic data source calls depending on the filetype. Markdown can have front matter, and HTML can have meta/OpenGraph tags.

I've implemented [a function to try to detect markdown](https://github.com/timelinize/timelinize/blob/8f0f491ad5ee322b1974388c1a584e49c060f8c5/timeline/processor.go#L478), but it's a bit naive; I have no idea how accurate/effective it is (probably not very accurate). File extension is a better bet, if possible.

**Comment by rubiojr** [comment]
Sounds good to me.

> I've implemented [a function to try to detect markdown](https://github.com/timelinize/timelinize/blob/8f0f491ad5ee322b1974388c1a584e49c060f8c5/timeline/processor.go#L478), but it's a bit naive; I have no idea how accurate/effective it is (probably not very accurate). File extension is a better bet, if possible.

We could keep an eye on https://github.com/google/magika/issues/96. It'll probably be significantly slower than file ext checking, and may complicate the 3rd party dependencies even further if it requires cgo, but we can start small with file ext and see how that plays out.

**Comment by rubiojr** [comment]
Related to file type detection, probably worth it's own issue, but I'll just drop a link to what we use in house and an existing Go port, that explains the heuristics nicely:

* https://github.com/github-linguist/linguist
* https://github.com/go-enry/go-enry

**Comment by mholt** [comment]
Thanks for the links!

I'd be open to exploring better type detection in the future, but yeah, speed and cgo complications are both concerns. So I'm glad you dropped the links here, and we can refer back to them if we're ready to upgrade our type detection. :+1: (Magika looks really cool... and we love Linguist of course!)
